data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java

```diff
@@ -0,0 +1,220 @@
+            this.startTime = startTime;
+            return this;
+        }
+
+        public Builder waitForExport(boolean waitForExport) {
+            this.waitForExport = waitForExport;
+            return this;
+        }
+
+        public ShardConsumer build() {
+            return new ShardConsumer(this);
+        }
+
+    }
+
+
+    @Override
+    public void run() {
+        LOG.debug("Shard Consumer start to run...");
+
+        long lastCheckpointTime = System.currentTimeMillis();
+        String sequenceNumber = "";
+
+        while (!shouldStop) {
+            if (shardIterator == null) {
+                // End of Shard
+                LOG.debug("Reach end of shard");
+                checkpointer.checkpoint(sequenceNumber);
+                break;
+            }
+
+            if (System.currentTimeMillis() - lastCheckpointTime > DEFAULT_CHECKPOINT_INTERVAL_MILLS) {
+                LOG.debug("Perform regular checkpointing for Shard Consumer");
+                checkpointer.checkpoint(sequenceNumber);
+                lastCheckpointTime = System.currentTimeMillis();
+            }
+
+            // Use the shard iterator to read the stream records
+            GetRecordsRequest req = GetRecordsRequest.builder()
+                    .shardIterator(shardIterator)
+                    .limit(MAX_GET_RECORD_ITEM_COUNT)
+                    .build();
+
+
+            List<Record> records;
+            GetRecordsResponse response;
+            try {
+                response = dynamoDbStreamsClient.getRecords(req);
+            } catch (SdkException e) {
+                checkpointer.checkpoint(sequenceNumber);
+                throw e;
+            }
+
+            shardIterator = response.nextShardIterator();
+
+            if (!response.records().isEmpty()) {
+                // Always use the last sequence number for checkpoint
+                sequenceNumber = response.records().get(response.records().size() - 1).dynamodb().sequenceNumber();
+
+                if (waitForExport) {
+                    Instant lastEventTime = response.records().get(response.records().size() - 1).dynamodb().approximateCreationDateTime();
+                    if (lastEventTime.compareTo(startTime) <= 0) {
+                        LOG.debug("Get {} events before start time, ignore...", response.records().size());
+                        continue;
+                    }
+                    checkpointer.checkpoint(sequenceNumber);
+                    waitForExport();
+                    waitForExport = false;
+
+                    records = response.records().stream()
+                            .filter(record -> record.dynamodb().approximateCreationDateTime().compareTo(startTime) > 0)
+                            .collect(Collectors.toList());
+                } else {
+                    records = response.records();
+                }
+                recordConverter.writeToBuffer(records);
```

Comment:
Why do both the file reader thread path and stream scheduler thread path write to the buffer? What is the difference in these outputs?

Reply:
They are converted into the same Jackson events and then sent to the same buffer and eventually same destination.